pieforproviders/pieforproviders
# [PIE-712] Earned revenue from dashboard

## 💅🏼 What issue does this fix?



Closes #712 

## 🍂 Before You Submit


## 🛷 Deployment


## 🧵 Steps to set up locally

```
bundle exec rails db:migrate
```

## 🧳 Steps to test
- create an accredited hourly Nebraska Rate in the LDDS region with an amount of 5.76
- create a business in one of the LDDS counties with a qris rating of step_five and make the business accredited
- create a child for that business (don't forget to add an active approval as well)
- set that child's active child approval to `special_needs_rate: false`
- create an attendance for that child during the active child approval period - let's say 3 hours and 21 minutes total_time_in_care
- the earned revenue for your created attendance should be 3.5 (total time in care rounded up to nearest 15 mins) * 5.76 (rate amount you set above) * 1.1576 (hardcoded step_five qris bump)

You can run a bunch of other manual tests, turning `special_needs_rate` on (and setting `special_needs_hourly_rate` or `special_needs_daily_rate`), turning accreditation off, setting the business' qris rating to `not_rated`, `step_one`, etc.

The best way to end-to-end test is to take ONE child from our current NE prediction algorithms spreadsheet, replicate them in your local, and replicate their attendances in your local, and then check to see if what you have in your dashboard for that child (after turning on the nebraska live algorithms feature flag) matches what the spreadsheet has for that child.

## 🕯 Caveats, concerns
This is the first time we're including rates, and it's based on several assumptions that are outlined in our #data slack channel, so if any of those are wrong we'll need to iterate.

Here are my "steps to test" for one of the kids in our prod system:

```
Steps to Test:
- reset & reseed db
- child = Child.find_by(full_name: "Rhonan Shaw")

Create Rates:

bundle exec rails nebraska_rates20210813

business changes:
accredited: false
step_two
county: Buffalo

Child record changes:
- change birthdate to match Client #89646689
- set enrolled_in_school to false

Approval record changes:
- Effective on: 2021-03-01
- Expires on: 2022-02-28

ChildApproval record changes:
- Special Needs Rate = false
- Special Needs Daily Rate = nil
- Special Needs Hourly Rate = nil
- authorized weekly hours = 45

NebraskaApprovalAmount record changes:
- Approval #1 - Family Fee	Approval #1 - Begin Date	Approval #1 - End Date	Approval #1 - Allocated Family Fee
158.00	2021-03-01	2022-02-28	79.00

Schedule:
2021-03-01	18:00:00		05:00:00	5
2021-03-01	18:00:00		05:00:00	4
2021-03-01	18:00:00		05:00:00	3
2021-03-01	18:00:00		05:00:00	2
2021-03-01	18:00:00		05:00:00	1

Attendances and Absences:

Attendance.create!(child_approval: ca, check_in: "2021-08-02 10:30:56+00")
Attendance.create!(child_approval: ca, check_in: "2021-08-03", absence: "absence")
Attendance.create!(child_approval: ca, check_in: "2021-08-04 15:45:54+00")
Attendance.create!(child_approval: ca, check_in: "2021-08-05 09:49:54+00", check_out: "2021-08-05 20:30:02+00")
Attendance.create!(child_approval: ca, check_in: "2021-08-06 10:31:57+00", check_out: "2021-08-06 21:30:33+00")
Attendance.create!(child_approval: ca, check_in: "2021-08-07 10:30:44+00", check_out: "2021-08-07 21:30:22+00")
Attendance.create!(child_approval: ca, check_in: "2021-08-09 10:30:58+00", check_out: "2021-08-09 21:57:11+00")
Attendance.create!(child_approval: ca, check_in: "2021-08-10 10:00:34+00", check_out: "2021-08-10 21:31:11+00")
Attendance.create!(child_approval: ca, check_in: "2021-08-11 10:59:02+00")
Attendance.create!(child_approval: ca, check_in: "2021-08-12 10:30:52+00", check_out: "2021-08-12 21:30:43+00")
Attendance.create!(child_approval: ca, check_in: "2021-08-13 10:06:38+00")

Compare to the dashboard on the spreadsheet for 8/13
```
Keep in mind when testing, there will probably be more attendances for that child that will also need to be added.
